File: backend/app/routers/categories.py
"""Category CRUD endpoints."""

import uuid
from typing import Annotated

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.database import get_db
from app.middleware.auth import get_current_user
from app.models.user import User
from app.schemas.category import (
    CategoryCreate,
    CategoryReorderRequest,
    CategoryResponse,
    CategoryTreeNode,
    CategoryUpdate,
    SetTagsRequest,
)
from app.schemas.series import SeriesListResponse
from app.services import category_service

router = APIRouter(prefix="/api/categories", tags=["categories"])


@router.get("/tree", response_model=list[CategoryTreeNode])
async def get_category_tree(
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Get the full category tree. Public endpoint for browsing."""
    return await category_service.get_category_tree(db)


@router.get("", response_model=list[CategoryResponse])
async def list_categories(
    _user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """List all categories flat."""
    return await category_service.list_categories(db)


@router.post("", response_model=CategoryResponse, status_code=status.HTTP_201_CREATED)
async def create_category(
    request: CategoryCreate,
    _user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Create a new category."""
    try:
        category = await category_service.create_category(
            db,
            name=request.name,
            icon=request.icon,
            description=request.description,
            parent_id=request.parent_id,
            tag_ids=request.tag_ids if request.tag_ids else None,
            match_mode=request.match_mode,
        )
        return category
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))


@router.patch("/reorder", status_code=status.HTTP_204_NO_CONTENT)
async def reorder_categories(
    request: CategoryReorderRequest,
    _user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Batch update sort_order for categories."""
    items = [{"id": item.id, "sort_order": item.sort_order} for item in request.items]
    await category_service.reorder_categories(db, items)


@router.get("/{category_id}", response_model=CategoryResponse)
async def get_category(
    category_id: uuid.UUID,
    _user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Get a single category with tags and subcategories."""
    category = await category_service.get_category(db, category_id)
    if category is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Category not found")
    return category


@router.patch("/{category_id}", response_model=CategoryResponse)
async def update_category(
    category_id: uuid.UUID,
    request: CategoryUpdate,
    _user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Update a category."""
    update_data = request.model_dump(exclude_unset=True)
    try:
        category = await category_service.update_category(db, category_id, **update_data)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    if category is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Category not found")
    return category


@router.delete("/{category_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_category(
    category_id: uuid.UUID,
    _user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Delete a category. Fails if it has subcategories."""
    try:
        found = await category_service.delete_category(db, category_id)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(e))
    if not found:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Category not found")


@router.put("/{category_id}/tags", response_model=CategoryResponse)
async def set_category_tags(
    category_id: uuid.UUID,
    request: SetTagsRequest,
    _user: Annotated[User, Depends(get_current_user)],
    db: Annotated[AsyncSession, Depends(get_db)],
):
    """Replace all tags for a category."""
    try:
        category = await category_service.set_category_tags(db, category_id, request.tag_ids)
    except ValueError as e:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST, detail=str(e))
    if category is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Category not found")
    return category


@router.get("/{category_id}/series", response_model=SeriesListResponse)
async def get_category_series(
    category_id: uuid.UUID,
    db: Annotated[AsyncSession, Depends(get_db)],
    page: int = Query(1, ge=1),
    per_page: int = Query(20, ge=1, le=100),
):
    """Get series under a category. Public endpoint."""
    result = await category_service.get_category_series(db, category_id, page, per_page)
    return result

File: frontend/src/stores/categories.ts
import { ref, computed } from 'vue'
import { defineStore } from 'pinia'
import axios from 'axios'
import api from '@/lib/api'
import type {
  Category,
  CategoryTreeNode,
  CategoryCreate,
  CategoryUpdate,
  ReorderItem,
  SeriesListResponse,
} from '@/types/category'

function extractError(e: unknown, fallback: string): string {
  if (axios.isAxiosError(e)) return e.response?.data?.detail ?? fallback
  return fallback
}

export const useCategoryStore = defineStore('categories', () => {
  // State
  const categories = ref<Category[]>([])
  const tree = ref<CategoryTreeNode[]>([])
  const currentCategory = ref<Category | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  // Getters
  const rootCategories = computed(() => categories.value.filter(c => c.parent_id === null))
  const getCategoryById = computed(() => {
    return (id: string) => categories.value.find(c => c.id === id) || null
  })

  // Actions
  async function fetchCategories() {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.get<Category[]>('/api/categories')
      categories.value = data
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to fetch categories')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function fetchTree() {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.get<CategoryTreeNode[]>('/api/categories/tree')
      tree.value = data
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to fetch category tree')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function fetchCategory(id: string) {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.get<Category>(`/api/categories/${id}`)
      currentCategory.value = data
      return data
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to fetch category')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function createCategory(payload: CategoryCreate) {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.post<Category>('/api/categories', payload)
      categories.value.push(data)
      return data
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to create category')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function updateCategory(id: string, payload: CategoryUpdate) {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.patch<Category>(`/api/categories/${id}`, payload)
      const index = categories.value.findIndex(c => c.id === id)
      if (index !== -1) categories.value[index] = data
      return data
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to update category')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function deleteCategory(id: string) {
    loading.value = true
    error.value = null
    try {
      await api.delete(`/api/categories/${id}`)
      categories.value = categories.value.filter(c => c.id !== id)
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to delete category')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function setCategoryTags(categoryId: string, tagIds: string[]) {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.put<Category>(`/api/categories/${categoryId}/tags`, { tag_ids: tagIds })
      const index = categories.value.findIndex(c => c.id === categoryId)
      if (index !== -1) categories.value[index] = data
      return data
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to set category tags')
      throw e
    } finally {
      loading.value = false
    }
  }

  async function reorderCategories(items: ReorderItem[]) {
    error.value = null
    try {
      await api.patch('/api/categories/reorder', { items })
      // Update local state
      for (const item of items) {
        const cat = categories.value.find(c => c.id === item.id)
        if (cat) cat.sort_order = item.sort_order
      }
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to reorder categories')
      throw e
    }
  }

  async function fetchCategorySeries(categoryId: string, page = 1, perPage = 20): Promise<SeriesListResponse> {
    loading.value = true
    error.value = null
    try {
      const { data } = await api.get<SeriesListResponse>(`/api/categories/${categoryId}/series`, {
        params: { page, per_page: perPage },
      })
      return data
    } catch (e: unknown) {
      error.value = extractError(e, 'Failed to fetch category series')
      throw e
    } finally {
      loading.value = false
    }
  }

  return {
    // State
    categories,
    tree,
    currentCategory,
    loading,
    error,
    // Getters
    rootCategories,
    getCategoryById,
    // Actions
    fetchCategories,
    fetchTree,
    fetchCategory,
    createCategory,
    updateCategory,
    deleteCategory,
    setCategoryTags,
    reorderCategories,
    fetchCategorySeries,
  }
})